### DISCUSSION TITLE
h11._util.LocalProtocolError: Too much data for declared Content-Length - when returning dict as response, with custom status code

### DISCUSSION THREAD

**Discussion by David-Lor** [fastapi/fastapi#7187] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI
from fastapi.responses import JSONResponse
from pydantic import BaseModel
import uvicorn

app = FastAPI()

class MyModel(BaseModel):
    foo: str
    bar: int

RESPONSE = {"message": "Hello there!"}

@app.get("/get1")
def get1():
    return RESPONSE

@app.get("/get2", status_code=204)
def get2():
    return RESPONSE

@app.post("/post1")
def post1():
    return RESPONSE

@app.post("/post2")
def post2(data: MyModel):
    return RESPONSE

@app.post("/post3", status_code=204)
def post3():
    return RESPONSE

@app.post("/post4", status_code=204)
def post4(data: MyModel):
    return RESPONSE

@app.post("/post5")
def post5(data: MyModel, status_code=204):
    return JSONResponse(
        content=RESPONSE,
        status_code=204 # must define status_code here
    )

if __name__ == "__main__":
    uvicorn.run(app, port=5000)
```

### Description

On the given code, the endpoints `/get2`, `/post3` and `/post4` fail, raising this exception:

```
INFO:     127.0.0.1:46692 - "POST /post4 HTTP/1.1" 204 No Content
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py", line 389, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/fastapi/applications.py", line 179, in __call__
    await super().__call__(scope, receive, send)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/starlette/applications.py", line 111, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc from None
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc from None
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/starlette/routing.py", line 566, in __call__
    await route.handle(scope, receive, send)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/starlette/routing.py", line 227, in handle
    await self.app(scope, receive, send)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/starlette/routing.py", line 44, in app
    await response(scope, receive, send)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/starlette/responses.py", line 139, in __call__
    await send({"type": "http.response.body", "body": self.body})
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/starlette/exceptions.py", line 68, in sender
    await send(message)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/starlette/middleware/errors.py", line 156, in _send
    await send(message)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py", line 483, in send
    output = self.conn.send(event)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/h11/_connection.py", line 469, in send
    data_list = self.send_with_data_passthrough(event)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/h11/_connection.py", line 502, in send_with_data_passthrough
    writer(event, data_list.append)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/h11/_writers.py", line 78, in __call__
    self.send_data(event.data, write)
  File "/home/david/.miniconda3/envs/fastapi+pydantic/lib/python3.8/site-packages/h11/_writers.py", line 98, in send_data
    raise LocalProtocolError("Too much data for declared Content-Length")
h11._util.LocalProtocolError: Too much data for declared Content-Length
```

It seems that this fails when returning a dict as response (without using JSONResponse), AND setting a status_code for the endpoint.

### Environment

* OS: Ubuntu 18.04
* Python version: 3.8.5
* FastAPI version: 0.61.1
* Uvicorn version: 0.12.2

**Comment by SirTelemak** [comment]
HTTP 204 No Content cannot contain a message body: https://tools.ietf.org/html/rfc7230#section-3.3.3
"Any response to a HEAD request and any response with a 1xx (Informational), 204 (No Content), or 304 (Not Modified) status code is always terminated by the first empty line after the  header fields, regardless of the header fields present in the message, and thus cannot contain a message body."

content-length forced in h11 to be equal to 0 (`h11._connection.py:75`) so when you return something in this endpoints there is exception raised

**Comment by SirTelemak** [comment]
To make valid endpoint that returns 204 status code you should use Response object:
```python3
@app.post("/post5")
def post5(data: MyModel, status_code=HTTPStatus.NO_CONTENT):
    return Response(status_code=HTTPStatus.NO_CONTENT.value)
```

**Comment by David-Lor** [comment]
@SirTelemak Thanks for the clarification! Big facepalm for me haha :facepalm:

**Comment by marcospb19** [comment]
I wish we could have a clearer error message for this to help debugging.

**Comment by r-owen** [comment]
It would really help to have two things:
* An example of a standard delete router (one that returns 204) in the docs. I could find no example of a "delete" router at all.
* Allow returning None, when decorated as `@router.delete("/messages/{id}", status_code=204)`. In other words treat None as "no data" (at least when the response code suggests it, but what else would None mean?). This is how I originally wrote my delete router. For some reason that worked fine with the /docs interactive UI (with the deployment) and also (of course) my unit tests. I only realized something was wrong when users could not delete using `requests` or `httpx` to communicate with the deployment.

Changing my delete router to return `fastapi.Response` did fix the problem, as suggested above, so I am very grateful for this bug report.

But I am really puzzled why my old way of doing things worked at all. I would have expected the interactive docs/ UI and my unit tests to both fail.

**Comment by geekgao** [comment]
` return Response(status_code=HTTPStatus.NO_CONTENT.value)`  
That's seems bad. Not simple.  @SirTelemak 
You can do like this:

```
@app.post("/post5", responses={204: {"model": None}})
def post5(data: MyModel):
    pass
```

**Comment by SirTelemak** [comment]
> That's seems bad. Not simple. 

Yea, declaring custom response model seems easier, ain't thought about it. Just hardcoded status code looks little wrong for me :)

**Comment by NucleaPeon** [comment]
I discovered this issue when trying to get a 304 response working. 

If this helps anyone else:
`fastapi.status` doesn't have a status equivalent like it does for 204. In the route decorator, I added `responses: {304: {}}` and then a `return Response(status_code=304)`. 
Maybe this is obvious to some, but it may help others like myself.

**Comment by Kludex** [comment, marked as answer]
This issue is already solved on the latest FastAPI version.

**Reply by r-owen** [reply]
What is the resolution? (What is now the recommended way to return 204)?

**Reply by Kludex** [reply]
```py
from fastapi import FastAPI

app = FastAPI()

@app.get("/", status_code=204)
def home():
    ...
```
This is enough. As I said, the issue on the description was fixed.

**Reply by r-owen** [reply]
And then return `fastapi.Response(status_code=204)` or raise `raise HTTPException(204, ...)`?

**Reply by Kludex** [reply]
You don't need to return anything... You can be explicit, and write `return None` if you want.

**Reply by r-owen** [reply]
I confess I'm still confused. I need to return 200 if the data is available (the normal case). The question is what to do if requested data does not exist. My code currently raises 404. That works fine. But my impression is that it is considered more normal to return 204 instead of 404. So I had hoped to find an easy way to do that. But status_code only allows specifying a single value, not both 200 and 204. The documentation offers the example of return 201 (CREATED) instead of 200.
